smartsim/_core/mli/infrastructure/comm/broadcaster.py

```diff
@@ -0,0 +1,238 @@
+        """Pop the next event to be sent from the queue.
+
+        :returns: The next event to send if any events are enqueued, otherwise `None`.
+        """
+        try:
+            return self._event_buffer.popleft()
+        except IndexError:
+            logger.debug(f"Broadcast buffer exhausted for {self.name}")
+
+        return None
+
+    def _broadcast(self, timeout: float = 0.001) -> BroadcastResult:
+        """Broadcasts all buffered events to registered event consumers.
+
+        :param timeout: Maximum time to wait (in seconds) for messages to send
+        :returns: BroadcastResult containing the number of messages that were
+        successfully and unsuccessfully sent for all consumers
+        :raises SmartSimError: If the channel fails to attach
+        :raises SmartSimError: If broadcasting fails
+        """
+        # allow descriptors to be empty since events are buffered
+        self._descriptors = set(x for x in self._backbone.notification_channels if x)
+        if not self._descriptors:
+            msg = f"No event consumers are registered for {self.name}"
+            logger.warning(msg)
+            return BroadcastResult(0, 0)
+
+        self._prune_unused_consumers()
+        self._log_broadcast_start()
+
+        num_listeners = len(self._descriptors)
+        num_sent = 0
+        num_failures = 0
+
+        # send each event to every consumer
+        while event := self._get_next_event():
+            logger.debug(f"Broadcasting {event=} to {num_listeners} listeners")
+            event_bytes = bytes(event)
+
+            for i, descriptor in enumerate(self._descriptors):
+                comm_channel = self._get_comm_channel(descriptor)
+
+                try:
+                    comm_channel.send(event_bytes, timeout)
+                    num_sent += 1
+                except Exception:
+                    msg = (
+                        f"Broadcast {i+1}/{num_listeners} for event {event.uid} to "
+                        f"channel {descriptor} from {self.name} failed."
+                    )
+                    logger.exception(msg)
+                    num_failures += 1
+
+        return BroadcastResult(num_sent, num_failures)
+
+    def send(self, event: EventBase, timeout: float = 0.001) -> int:
+        """Implementation of `send` method of the `EventPublisher` protocol. Publishes
+        the supplied event to all registered broadcast consumers.
+
+        :param event: An event to publish
+        :param timeout: Maximum time to wait (in seconds) for messages to send
+        :returns: The total number of events successfully published to consumers
+        :raises ValueError: If event serialization fails
+        :raises AttributeError: If event cannot be serialized
+        :raises KeyError: If channel fails to attach using registered descriptors
+        :raises SmartSimError: If any unexpected error occurs during send
+        """
+        try:
+            self._save_to_buffer(event)
```

Comment:
Here in send you mention SmartSimError once when it is raised twice, but in _broadcast docstring you mention it twice

Reply:
converted `_broadcast` to a single :raises: docstring